Docs: Add recommendations about Session usage.

Lorak-mmk opened the pull request:
#216 showed that best practices of `Session` usage are not obvious and should be stated in documentation.
This PR adds the relevant section that explains how to use it in order to not decrease performance.

Fixes: #216 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [ ] I have split my patch into logically separate commits.
- [ ] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [ ] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

piodul commented on docs/source/connecting/connecting.md:
```diff
@@ -35,6 +35,12 @@ async fn main() -> Result<(), Box<dyn Error>> {
+Driver maintains this pool of connections and each connection is capable of handling multiple requests in parallel. Driver will also route the request to node / shard that actually owns the data (unless load balancing policy that you use doesn't support it).
+For those reasons, we recommend to use one instance of `Session` per application.
+Creating short-lived `Session`'s (e.g. `Session` per request) is extremely discouraged and will result in great performance penalties because creating a `Session` is a costly process - it requires estabilishing a lot of TCP connections.
+Creating many `Session`'s in one application (e.g. `Session` per thread / per Tokio taks) is also discouraged, because it wastes resources - as mentioned before, `Session` maintains connection pool itself and can handle parallel queries, so you would be holding a lot of connections unnecessarily.
```
typo: `per Tokio task`

piodul replied:
`maintains connection pool` -> `maintains a connection pool`

piodul commented on docs/source/connecting/connecting.md:
```diff
@@ -35,6 +35,12 @@ async fn main() -> Result<(), Box<dyn Error>> {
+For those reasons, we recommend to use one instance of `Session` per application.
+Creating short-lived `Session`'s (e.g. `Session` per request) is extremely discouraged and will result in great performance penalties because creating a `Session` is a costly process - it requires estabilishing a lot of TCP connections.
+Creating many `Session`'s in one application (e.g. `Session` per thread / per Tokio taks) is also discouraged, because it wastes resources - as mentioned before, `Session` maintains connection pool itself and can handle parallel queries, so you would be holding a lot of connections unnecessarily.
+If you need to share `Session` to different threads / Tokio tasks etc. use `Arc<Session>` - all methods of `Session` take `&self`, so it doesn't hinder the functionality in any way. In order to avoid cost of `Arc`'s atomic operations, you can estabilish all your `Arc`'s in the beggining as the atomic operations only happend when cloning `Arc`, not when using it's inner object.
```
`share 'Session' to different threads` -> `share 'Session' with different threads`

piodul replied:
`avoid cost` -> `avoid the cost`

piodul replied:
typo: `happend` -> `happen`

piodul replied:
`it's inner object` -> `its inner object`

piodul replied:
I'm not sure that the remark about avoiding `Arc` atomic operations is realistic. Many client programs will spawn new tasks (I bet most of the web frameworks based on tokio do this) and then you will have to clone the `Arc`. The only alternative I see is putting the session into a global - but then you would either have to use unsafe to access it, or some wrapper that introduces synchronization overhead (rwlock, once cell, lazy static etc.)

Lorak-mmk replied:
Dropped this section

piodul commented on docs/source/connecting/connecting.md:
```diff
@@ -35,6 +35,12 @@ async fn main() -> Result<(), Box<dyn Error>> {
 After successfully connecting to some specified node the driver will fetch topology information about
 other nodes in this cluster and connect to them as well.
 
+Driver maintains this pool of connections and each connection is capable of handling multiple requests in parallel. Driver will also route the request to node / shard that actually owns the data (unless load balancing policy that you use doesn't support it).
```
`Driver maintains this pool`? Maybe `a pool`, or `its own pool`?

piodul replied:
Also, `The driver` instead of `Driver` sounds slightly better

piodul replied:
`load balancing policy` -> `the load balancing policy`

piodul replied:
`the request` -> `requests`?

Lorak-mmk replied:
> `the request` -> `requests`?

Changed to `Driver will also route requests to nodes / shards that actually own the data (unless the load balancing policy that you use doesn't support it).`

Lorak-mmk replied:
> `Driver maintains this pool`? Maybe `a pool`, or `its own pool`?

I was referring to a pool that was mentioned in a sentence before. But now that I think of it your version may be easier to understand.
Changed to `The driver maintains its own pool of connections to each node and each connection is capable of handling multiple requests in parallel.`

piodul commented on docs/source/connecting/connecting.md:
```diff
@@ -35,6 +35,12 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
+Driver maintains this pool of connections and each connection is capable of handling multiple requests in parallel. Driver will also route the request to node / shard that actually owns the data (unless load balancing policy that you use doesn't support it).
+For those reasons, we recommend to use one instance of `Session` per application.
+Creating short-lived `Session`'s (e.g. `Session` per request) is extremely discouraged and will result in great performance penalties because creating a `Session` is a costly process - it requires estabilishing a lot of TCP connections.
```
IMO "extremely" is too strong of a word here. There are cases where short sessions make sense, e.g. an administrative tool that connects only to execute a single statement. Of course, creating multiple sessions over the course of the application is wasteful.

Lorak-mmk replied:
strongly?

Lorak-mmk replied:
I change to `strongly`, if that's also too strong I'll rewrite this sentence differently

piodul replied:
"Strongly" sounds better to me as well. I was mostly concerned about the "short-lived" sessions given the use cases for them I mentioned in my previous comment, but with the added warning about using only one session per application I think it looks OK now.

piodul commented on docs/source/connecting/connecting.md:
```diff
@@ -35,6 +35,12 @@ async fn main() -> Result<(), Box<dyn Error>> {
 other nodes in this cluster and connect to them as well.
 
+Driver maintains this pool of connections and each connection is capable of handling multiple requests in parallel. Driver will also route the request to node / shard that actually owns the data (unless load balancing policy that you use doesn't support it).
+For those reasons, we recommend to use one instance of `Session` per application.
```
`recommend to use` -> `recommend using`

Lorak-mmk commented:
v2: 
- Addressed review comments
- Changed formatting so that new section doesn't render as a block of text
- Added an adminition in order to make this section easier to notice. How it looks in Scylla docs:
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/5740e1a3-b7b3-46d5-b8d5-1919b8d9b4fd)
and in mdbook:
![image](https://github.com/scylladb/scylla-rust-driver/assets/5958117/99d83023-dcda-42d9-b1a2-916bccf8f8bf)
This required changing preprocessor script a bit.
- Split the file into sections becaue there was too much contiguous text

Lorak-mmk commented:
v3: Fixed typo in preprocessor script
